File: source/ios/NativeGATracker/Commands/TrackTransaction.m
/**
 * Project: ANE-Google-Analytics
 *
 * Author:  Alessandro Bianco
 * Website: http://alessandrobianco.eu
 * Twitter: @alebianco
 * Created: 10/09/13 14:04
 *
 * Copyright © 2013 Alessandro Bianco
 */

#import "TrackTransaction.h"

@implementation TrackTransaction

-(FREObject) track:(FREObject *)data onTracker:(id)tracker forContext:(FREContext)context {

    FREObject result = NULL;

    NSString *id;
    NSNumber *revenue;
    NSString *affiliation;
    NSNumber *shipping;
    NSNumber *tax;
    NSString *currency;

    FREObject *products;

    @try {
        id = [FREConversionUtil toString:[FREConversionUtil getProperty:@"id" fromObject:data]];
    }
    @catch (NSException *exception) {
        logEvent(context, kError, @"Unable to read 'id' property. [Exception:(type:%@, method:%s)].", [exception name], __FUNCTION__);
        return createRuntimeException(@"ArgumentError", 0, @"Unable to read 'id' property on method '%s'.", __FUNCTION__);
    }
    
    @try {
        affiliation = [FREConversionUtil toString:[FREConversionUtil getProperty:@"affiliation" fromObject:data]];
    }
    @catch (NSException *exception) {
        logEvent(context, kError, @"Unable to read 'affiliation' property. [Exception:(type:%@, method:%s)].", [exception name], __FUNCTION__);
        return createRuntimeException(@"ArgumentError", 0, @"Unable to read 'affiliation' property on method '%s'.", __FUNCTION__);
    }
    
    @try {
        revenue = [FREConversionUtil toNumber:[FREConversionUtil getProperty:@"revenue" fromObject:data]];
    }
    @catch (NSException *exception) {
        logEvent(context, kError, @"Unable to read 'revenue' property. [Exception:(type:%@, method:%s)].", [exception name], __FUNCTION__);
        return createRuntimeException(@"ArgumentError", 0, @"Unable to read 'revenue' property on method '%s'.", __FUNCTION__);
    }
    
    @try {
        shipping = [FREConversionUtil toNumber:[FREConversionUtil getProperty:@"shipping" fromObject:data]];
    }
    @catch (NSException *exception) {
        logEvent(context, kError, @"Unable to read 'shipping' property. [Exception:(type:%@, method:%s)].", [exception name], __FUNCTION__);
        return createRuntimeException(@"ArgumentError", 0, @"Unable to read 'shipping' property on method '%s'.", __FUNCTION__);
    }
    
    @try {
        tax = [FREConversionUtil toNumber:[FREConversionUtil getProperty:@"tax" fromObject:data]];
    }
    @catch (NSException *exception) {
        logEvent(context, kError, @"Unable to read 'tax' property. [Exception:(type:%@, method:%s)].", [exception name], __FUNCTION__);
        return createRuntimeException(@"ArgumentError", 0, @"Unable to read 'tax' property on method '%s'.", __FUNCTION__);
    }
    
    @try {
        currency = [FREConversionUtil toString:[FREConversionUtil getProperty:@"currency" fromObject:data]];
    }
    @catch (NSException *exception) {
        logEvent(context, kInfo, @"Unable to read 'currency' property, falling back to default label. [Exception:(type:%@, method:%s)].", [exception name], __FUNCTION__);
        currency = NULL;
    }
    
    NSMutableDictionary *hit = [NSDictionary dictionaryWithObjectsAndKeys:
                                kGAITransaction, kGAIHitType,
                                id, kGAITransactionId,
                                affiliation, kGAITransactionAffiliation,
                                [revenue stringValue], kGAITransactionRevenue,
                                [shipping stringValue], kGAITransactionShipping,
                                [tax stringValue], kGAITransactionTax,
                                nil];
    if (currency) {
        [hit setObject:currency forKey:kGAICurrencyCode];
    }
    
    [tracker send:hit];
    
    @try {
        Track *productTrackCommand = (Track *)[[TrackProduct alloc] init];
        products = [FREConversionUtil getProperty:@"products" fromObject:data];
        NSUInteger prodc = [FREConversionUtil getArrayLength:products];
        for (NSUInteger i = 0; i < prodc; i++) {
            FREObject product = [FREConversionUtil getArrayItemAt:i on:products];
            result = [productTrackCommand track:product onTracker:tracker forContext:context];
            if (result != NULL) {
                return result;
            }
        }
    }
    @catch (NSException *exception) {
        logEvent(context, kError, @"Unable to read a 'products' property. [Exception:(type:%@, method:%s)].", [exception name], __FUNCTION__);
        return createRuntimeException(@"ArgumentError", 0, @"Unable to read a 'products' property '%s'.", __FUNCTION__);
    }

    return result;
}

@end

File: source/android/src/eu/alebianco/air/extensions/analytics/commands/TrackTransaction.java
package eu.alebianco.air.extensions.analytics.commands;

import com.adobe.fre.*;
import com.google.analytics.tracking.android.Fields;
import com.google.analytics.tracking.android.HitTypes;
import com.google.analytics.tracking.android.Tracker;
import com.stackoverflow.util.StackTraceInfo;
import eu.alebianco.air.extensions.utils.FREUtils;
import eu.alebianco.air.extensions.utils.LogLevel;

import java.util.HashMap;
import java.util.Map;

/**
 * Project: ANE-Google-Analytics
 *
 * Author:  Alessandro Bianco
 * Website: http://alessandrobianco.eu
 * Twitter: @alebianco
 * Created: 10/09/2013 09:39
 *
 * Copyright © 2013 Alessandro Bianco
 */
public class TrackTransaction implements ITrackHit {
    @Override
    public FREObject track(FREContext context, Tracker tracker, FREObject data) {

        FREObject result;

        String id;
        Double revenue;
        String affiliation;
        Double shipping;
        Double tax;
        String currency;

        FREArray products;

        try {
            id = data.getProperty("id").getAsString();
        } catch (Exception e) {
            FREUtils.logEvent(context, LogLevel.ERROR,
                    "Unable to read 'id' property. (Exception:[name:%s, reason:%s, method:%s:%s])",
                    FREUtils.stripPackageFromClassName(e.toString()), FREUtils.stripPackageFromClassName(StackTraceInfo.getCurrentClassName()), FREUtils.stripPackageFromClassName(StackTraceInfo.getCurrentMethodName()));
            return FREUtils.createRuntimeException("ArgumentError", 0, "Unable to read 'id' property on method '%s:%s'.", FREUtils.stripPackageFromClassName(StackTraceInfo.getCurrentClassName()), FREUtils.stripPackageFromClassName(StackTraceInfo.getCurrentMethodName()));
        }

        try {
            revenue = data.getProperty("revenue").getAsDouble();
        } catch (Exception e) {
            FREUtils.logEvent(context, LogLevel.ERROR,
                    "Unable to read 'revenue' property. (Exception:[name:%s, reason:%s, method:%s:%s])",
                    FREUtils.stripPackageFromClassName(e.toString()), FREUtils.stripPackageFromClassName(StackTraceInfo.getCurrentClassName()), FREUtils.stripPackageFromClassName(StackTraceInfo.getCurrentMethodName()));
            return FREUtils.createRuntimeException("ArgumentError", 0, "Unable to read 'revenue' property on method '%s:%s'.", FREUtils.stripPackageFromClassName(StackTraceInfo.getCurrentClassName()), FREUtils.stripPackageFromClassName(StackTraceInfo.getCurrentMethodName()));
        }

        try {
            affiliation = data.getProperty("affiliation").getAsString();
        } catch (Exception e) {
            FREUtils.logEvent(context, LogLevel.ERROR,
                    "Unable to read 'affiliation' property. (Exception:[name:%s, reason:%s, method:%s:%s])",
                    FREUtils.stripPackageFromClassName(e.toString()), FREUtils.stripPackageFromClassName(StackTraceInfo.getCurrentClassName()), FREUtils.stripPackageFromClassName(StackTraceInfo.getCurrentMethodName()));
            return FREUtils.createRuntimeException("ArgumentError", 0, "Unable to read 'affiliation' property on method '%s:%s'.", FREUtils.stripPackageFromClassName(StackTraceInfo.getCurrentClassName()), FREUtils.stripPackageFromClassName(StackTraceInfo.getCurrentMethodName()));
        }

        try {
            shipping = data.getProperty("shipping").getAsDouble();
        } catch (Exception e) {
            FREUtils.logEvent(context, LogLevel.ERROR,
                    "Unable to read 'shipping' property. (Exception:[name:%s, reason:%s, method:%s:%s])",
                    FREUtils.stripPackageFromClassName(e.toString()), FREUtils.stripPackageFromClassName(StackTraceInfo.getCurrentClassName()), FREUtils.stripPackageFromClassName(StackTraceInfo.getCurrentMethodName()));
            return FREUtils.createRuntimeException("ArgumentError", 0, "Unable to read 'shipping' property on method '%s:%s'.", FREUtils.stripPackageFromClassName(StackTraceInfo.getCurrentClassName()), FREUtils.stripPackageFromClassName(StackTraceInfo.getCurrentMethodName()));
        }

        try {
            tax = data.getProperty("tax").getAsDouble();
        } catch (Exception e) {
            FREUtils.logEvent(context, LogLevel.ERROR,
                    "Unable to read 'tax' property. (Exception:[name:%s, reason:%s, method:%s:%s])",
                    FREUtils.stripPackageFromClassName(e.toString()), FREUtils.stripPackageFromClassName(StackTraceInfo.getCurrentClassName()), FREUtils.stripPackageFromClassName(StackTraceInfo.getCurrentMethodName()));
            return FREUtils.createRuntimeException("ArgumentError", 0, "Unable to read 'tax' property on method '%s:%s'.", FREUtils.stripPackageFromClassName(StackTraceInfo.getCurrentClassName()), FREUtils.stripPackageFromClassName(StackTraceInfo.getCurrentMethodName()));
        }

        try {
            currency = data.getProperty("currency").getAsString();
        } catch (Exception e) {
            FREUtils.logEvent(context, LogLevel.INFO,
                    "Unable to read 'currency' property, falling back to default value. (Exception:[name:%s, reason:%s, method:%s:%s])",
                    FREUtils.stripPackageFromClassName(e.toString()), FREUtils.stripPackageFromClassName(StackTraceInfo.getCurrentClassName()), FREUtils.stripPackageFromClassName(StackTraceInfo.getCurrentMethodName()));
            currency = null;
        }

        HashMap<String, String> hit = new HashMap<String, String>();
        hit.put(Fields.HIT_TYPE, HitTypes.TRANSACTION);
        hit.put(Fields.TRACKING_ID, id);
        hit.put(Fields.TRANSACTION_AFFILIATION, affiliation);
        hit.put(Fields.TRANSACTION_REVENUE, revenue.toString());
        hit.put(Fields.TRANSACTION_TAX, tax.toString());
        hit.put(Fields.TRANSACTION_SHIPPING, shipping.toString());
        if (currency != null) {
            hit.put(Fields.CURRENCY_CODE, currency);
        }

        tracker.send(hit);

        try {
            ITrackHit productTrackCommand = new TrackProduct();
            products = (FREArray) data.getProperty("products");
            long numProducts = products.getLength();
            for (long i = 0; i < numProducts; i++) {
                FREObject product = products.getObjectAt(i);
                result = productTrackCommand.track(context, tracker, product);
                if (result != null) {
                    return result;
                }
            }
        } catch (Exception e) {
            FREUtils.logEvent(context, LogLevel.ERROR,
                    "Unable to read a 'products' property. (Exception:[name:%s, reason:%s, method:%s:%s])",
                    FREUtils.stripPackageFromClassName(e.toString()), FREUtils.stripPackageFromClassName(StackTraceInfo.getCurrentClassName()), FREUtils.stripPackageFromClassName(StackTraceInfo.getCurrentMethodName()));
            return FREUtils.createRuntimeException("ArgumentError", 0, "Unable to read a 'products' property '%s:%s'.", FREUtils.stripPackageFromClassName(StackTraceInfo.getCurrentClassName()), FREUtils.stripPackageFromClassName(StackTraceInfo.getCurrentMethodName()));
        }

        return null;
    }
}
